### PR TITLE
refactor: add metrics endpoint (DEV-1555)

### PR DESCRIPTION
**Pull request by subotic**
### Task Description/Number

Issue Number: DEV-1555

### Basic Requirements

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

### PR Type

What kind of change does this PR introduce?

- [ ] Bugfix: represents bug fixes
- [x] Refactor: represents production code refactoring
- [ ] Feature: represents a new feature
- [ ] Documentation: documentation changes (no production code change)
- [ ] Chore: maintenance tasks (no production code change)
- [ ] Style: styles updates (no production code change)
- [ ] Test: all about tests: adding, refactoring tests (no production code change)
- [ ] Other... Please describe:

### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### Does this PR change client-test-data?

- [ ] Yes (don't forget to update the JS-LIB team about the change)
- [x] No

### CODE REVIEW

**Review by irinaschubert — APPROVED** [review]
LGTM

File: webapi/src/main/scala/org/knora/webapi/instrumentation/metrics/PrometheusServer.scala [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import zio.metrics.connectors.MetricsConfig
+import zio.metrics.connectors.prometheus
+import zio.metrics.connectors.prometheus.PrometheusPublisher
+import zio.metrics.jvm.DefaultJvmMetrics
+
+import org.knora.webapi.config.AppConfig
+
+object PrometheusServer {
+
+  private val nThreads = 5
```
Comment (BalduinLandolt) [inline webapi/src/main/scala/org/knora/webapi/instrumentation/metrics/PrometheusServer.scala:20]: why is this not coming from the config?

**Review by BalduinLandolt — COMMENTED** [review]
Looks good code-wise, I would like to have seen some documentation on it

File: webapi/src/main/scala/org/knora/webapi/Main.scala [hunk 2]
```diff
@@ -30,6 +31,10 @@ object Main extends ZIOApp {
   ] = ZLayer.empty ++ Runtime.removeDefaultLoggers ++ SLF4J.slf4j ++ LayersLive.dspLayersLive
 
   /* Here we start our Application */
-  override def run = AppServer.live.launch
-
+  override def run = for {
+    f1 <- PrometheusServer.make.forkDaemon
+    f2 <- (AppServer.live *> ZIO.never).forkDaemon
+    _  <- f1.join
+    _  <- f2.join
```
Comment (mpro7) [inline webapi/src/main/scala/org/knora/webapi/Main.scala:38]:
```suggestion
    _ <- PrometheusServer.make.forkDaemon.join
    _ <- (AppServer.live *> ZIO.never).forkDaemon.join
```

How about join in the same lines?

**Review by mpro7 — APPROVED** [review]
Looks good. Feel free to apply my suggestion. Shouldn't `PrometheusServer` has some unit tests written?

**Comment by seakayone** [comment]
As discussed.  I would prefer to have all zio http server definitions in the core package side by side.

I do think it will make sense to split the application and instrumentation http endpoint to be on separate port, hence the new Server should be named `InstrumentationHttpServer` or something like that .

Please also move the health route to the instrumentation server.

File: webapi/src/main/scala/org/knora/webapi/Main.scala [hunk 3]
```diff
@@ -30,6 +31,10 @@ object Main extends ZIOApp {
   ] = ZLayer.empty ++ Runtime.removeDefaultLoggers ++ SLF4J.slf4j ++ LayersLive.dspLayersLive
 
   /* Here we start our Application */
-  override def run = AppServer.live.launch
-
+  override def run = for {
+    f1 <- PrometheusServer.make.forkDaemon
+    f2 <- (AppServer.live *> ZIO.never).forkDaemon
```
[the remark this answers is not in the document]
Comment (subotic) [reply webapi/src/main/scala/org/knora/webapi/Main.scala:36]: `.launch` ist a method on `ZLayer`. It does exactly the same.

File: webapi/src/main/scala/org/knora/webapi/instrumentation/metrics/PrometheusServer.scala (hunk 1 above)
Comment (subotic) [reply webapi/src/main/scala/org/knora/webapi/instrumentation/metrics/PrometheusServer.scala:20]: I'm not sure what it does exactly. I still need to figure that out first. Maybe we don't need it.

**Comment by subotic** [comment]
> Looks good code-wise, I would like to have seen some documentation on it

Docs added.

**Comment by subotic** [comment]
> As discussed. I would prefer to have all zio http server definitions in the core package side by side.
> 
> I do think it will make sense to split the application and instrumentation http endpoint to be on separate port, hence the new Server should be named `InstrumentationHttpServer` or something like that .
> 
> Please also move the health route to the instrumentation server.

done.

**Pull request merged**